odilo.py

```diff
@@ -0,0 +1,598 @@
+        "MP3": Edition.AUDIO_MEDIUM,
+        "MP4": Edition.VIDEO_MEDIUM,
+        "WMV": Edition.VIDEO_MEDIUM,
+        "JPG": Edition.IMAGE_MEDIUM,
+        "SCORM": Edition.COURSEWARE_MEDIUM
+    }
+
+    @classmethod
+    def record_info_to_circulation(cls, availability):
+        """ Note:  The json data passed into this method is from a different file/stream
+        from the json data that goes into the record_info_to_metadata() method.
+        """
+
+        if 'recordId' not in availability:
+            return None
+
+        record_id = availability['recordId']
+        primary_identifier = IdentifierData(Identifier.ODILO_ID, record_id)  # We own this availability.
+
+        licenses_owned = int(availability['totalCopies'])
+        licenses_available = int(availability['availableCopies'])
+
+        # 'licenses_reserved' is the number of patrons who put the book on hold earlier,
+        #  but who are now at the front of the queue and who could get the book right now if they wanted to.
+        if 'notifiedHolds' in availability:
+            licenses_reserved = int(availability['notifiedHolds'])
+        else:
+            licenses_reserved = 0
+
+        # 'patrons_in_hold_queue' contains the number of patrons who are currently waiting for a copy of the book.
+        if 'holdsQueueSize' in availability:
+            patrons_in_hold_queue = int(availability['holdsQueueSize'])
+        else:
+            patrons_in_hold_queue = 0
+
+        return CirculationData(
+            data_source=DataSource.ODILO,
+            primary_identifier=primary_identifier,
+            licenses_owned=licenses_owned,
+            licenses_available=licenses_available,
+            licenses_reserved=licenses_reserved,
+            patrons_in_hold_queue=patrons_in_hold_queue,
+        )
+
+    @classmethod
+    def image_link_to_linkdata(cls, link, rel):
+        if not link:
+            return None
+
+        return LinkData(rel=rel, href=link, media_type=Representation.JPEG_MEDIA_TYPE)
+
+    @classmethod
+    def record_info_to_metadata(cls, book, availability):
+        """Turn Odilo's JSON representation of a book into a Metadata
+        object.
+
+        Note:  The json data passed into this method is from a different file/stream
+        from the json data that goes into the book_info_to_circulation() method.
+        """
+        if 'id' not in book:
+            return None
+
+        odilo_id = book['id']
+        primary_identifier = IdentifierData(Identifier.ODILO_ID, odilo_id)
+        active = book.get('active')
+
+        title = book.get('title')
+        subtitle = book.get('subtitle')
+        series = book.get('series')
+        series_position = book.get('seriesPosition')
+
+        contributors = []
+        author = book.get('author')
+        if author:
+            roles = [Contributor.AUTHOR_ROLE]
+            contributor = ContributorData(sort_name=author, display_name=author, roles=roles, biography=None)
```

Comment:
Treating `sort_name` and `display_name` is a problem. I see from your test that your names are in the format 'E. Veale'. We need names in the format 'Veale, E.' (that's `sort_name`), with 'E. Veale' (that's `display_name`) being an optional addition. If you just don't have the `sort_name` then we'll figure something out, but it would be best to modify your API to provide it.

Reply:
We use just "display_name"s, we don't have the "sort_name". I could add it to the API, but it would be "manually", as we don't have this data cataloged... or whatever other solution you suggest.

Reply:
OK, as I said this is a bit of a problem. We have a web API for guessing `sort_name` from `display_name`. You can activate this API by calling `MetadataWranglerOPDSLookup.from_config(_db, collection)` and then passing the resulting object into `Metadata.apply()` as the `metadata_client` argument.

Behind the scenes, this makes an API call like this:

http://metadata.librarysimplified.org/canonical-author-name?display_name=E.%20Veale&urn=urn:isbn:1234567890

It sends over the display_name and (if possible) the ISBN of the book. The metadata wrangler uses these two pieces of information to query OCLC and VIAF in hopes of finding the author's canonical name (e.g. "Veale, E" on http://viaf.org/viaf/49093834/).

The problem is that this process is quite slow and the metadata wrangler is already overloaded. In the long term this is the right solution but you won't be able to get it to work this week. In the short term, you can use the `display_name_to_sort_name` function defined in `core/util/personal_names.py`, which uses heuristics to convert "E. Veale" into "Veale, E." most of the time.

Reply:
Ok, so, by the moment I will use the ```display_name_to_sort_name``` function as you suggest me, thank you!